nntrainer/layers/layer_context.cpp

```diff
@@ -172,9 +172,6 @@ Tensor &RunLayerContext::getWeightGrad(unsigned int idx) const {
  */
 Tensor &RunLayerContext::getWeightOptVar(unsigned int idx,
                                          unsigned int jdx) const {
-  if (!weights[idx]->hasGradient())
```

Comment:
I wonder if it's safe to remove this check. If weight is not trainable, its optimizer variables would not be allocated memory.

Reply:
we need to check 1) weight index is in range, 2) weight opt index is in range. Having gradient or not is orthogonal in terms of this function. As other functions don't check bounds for performance reason, I think we can let this be. But I'll add a comment in the header to state that this function does not check any bounds.

Reply:
The issue is not the bounds, but if the tensor will be empty or allocated. 
Optimizer requests optimizer variables for all the weights even if they are trainable or not.
However, the optimizer requested tensor will not be allocated (in my understanding) if the weight is not trainable.
The lines in discussion performs the check if the weight is trainable or not by checking if a gradient was requested for the weight.